src/Type/Constant/ConstantArrayType.php

```diff
@@ -300,6 +300,10 @@ public function acceptsWithReason(Type $type, bool $strictTypes): AcceptsResult
 			return $type->isAcceptedWithReasonBy($this, $strictTypes);
 		}
 
+		if ($type instanceof NeverType) {
```

Comment:
Please try instanceof CompoundType. You may also remove the above if with MixedType.

Reply:
Unfortunately the best we can do right now is `$type instanceof CompoundType && !$type instanceof IntersectionType`, because accessor based intersection types currently don't support "constant" arrays / array shapes.
E.g. `array{foo: mixed}` will not be accepted by `array&hasOffset('foo')`

`ConstantArrayType` always needs to look at the sum of the intersection types in that regard. (or maybe I'm missing some simple adjustment I could do)

I spent some time looking into this, specifically adjusting `TypeCombinator::intersect()` to combine `array&hasOffset('test')` directly into `array{test: mixed}` etc.
There are quite a few cases to think through there, so it'll take me some more time for a proposal, which would be better suited for a separate PR anyway.
Maybe I'll look into https://github.com/phpstan/phpstan/issues/8438#issuecomment-1936621275 as part of this, especially because avoiding general array intersection along with it seems to make it extra cumbersome to implement.

Anyway, right now I have the issue that `instanceof IntersectionType` is not allowed. Is there a better way (or do you envision one) to check for an intersection without adding an entry to the baseline?

Reply:
Thank you for this analysis. How come you can dissect a problem so well? :) Did you spend a lot of time writing custom extensions or looking through the codebase before you started contributing? It's awesome how well you can solve problems here! I'm looking forward to more contributions from you.

I'm going to push an update to the baseline that makes the error go away, we're already ignoring it once in the same file.

Why `instanceof *Type` is discouraged is described here https://phpstan.org/blog/why-is-instanceof-type-wrong-and-getting-deprecated - I'd like to get rid of all instances of such code. It's relatively easy to do on the outskirts of the codebase (like in Rules etc.) but it's harder when you're deep in the typesystem or in TypeCombinator.

Every time I get rid of `instanceof *Type` somewhere, one or two bugs get squished.

Once we're able to get rid of all `instanceof *Type`, we'll be able to stop using inheritance in the typesystem - so for example, ConstantStringType won't longer extend StringType. This will untie our hands and for example make it easy to make any type subtractable. (Subtractable types: Right now for example `MixedType` is subtractable - we're able to tell "this is mixed but it's never an int").

Reply:
I do have quite a lot of experience with PHP and complex code bases, so questions like "how would I build this?" or "what would make sense here?" seem to often lead me into the right direction. :)
But mainly it's just a test case, a debugger and time :P (and having done some small bugfixes in the Psalm code base helped familiarize myself a bit with static analyzers in general)

Though I also do tend to skim through existing (merged) PRs and issues before contributing anywhere to familiarize myself with the expectations and opinions of the maintainers.